app/xfd/man/xfd.man
.\" Copyright 1989, 1994, 1998  The Open Group
.\"
.\" Permission to use, copy, modify, distribute, and sell this software and its
.\" documentation for any purpose is hereby granted without fee, provided that
.\" the above copyright notice appear in all copies and that both that
.\" copyright notice and this permission notice appear in supporting
.\" documentation.
.\"
.\" The above copyright notice and this permission notice shall be included
.\" in all copies or substantial portions of the Software.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
.\" OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
.\" MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
.\" IN NO EVENT SHALL THE OPEN GROUP BE LIABLE FOR ANY CLAIM, DAMAGES OR
.\" OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
.\" ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
.\" OTHER DEALINGS IN THE SOFTWARE.
.\"
.\" Except as contained in this notice, the name of The Open Group shall
.\" not be used in advertising or otherwise to promote the sale, use or
.\" other dealings in this Software without prior written authorization
.\" from The Open Group.
.\"
.\"
.TH XFD 1 2023-02-11 __xorgversion__
.SH NAME
xfd \- display all the characters in an X font
.SH SYNOPSIS
.B xfd
[\-options ...] \fB\-fn\fP \fIfontname\fP
.PP
.B xfd
[\-options ...] \fB\-fa\fP \fIfontname\fP
.SH DESCRIPTION
The \fBxfd\fP utility creates a window containing the name of the font being
displayed, a row of command buttons, several lines of text for displaying
character metrics, and a grid containing one glyph per cell.  The
characters are shown in increasing order from left to right, top to
bottom.  The first character displayed at the top left will be character
number 0 unless the \fB\-start\fP option has been supplied in which case the
character with the number given in the \fB\-start\fP option will be used.
.PP
The characters are displayed in a grid of boxes, each large enough to hold
any single character in the font.  Each character glyph is drawn using
the PolyText16 request (used by the \fIXlib\fP routine \fBXDrawString16\fP)
or the XftDrawString32 routine, depending on whether the \fB\-fn\fP or
\fB\-fa\fP option was given.
If the \fB\-box\fP option is given, a rectangle will be drawn around each
character, showing where an ImageText16 request (used by the \fIXlib\fP
routine \fBXDrawImageString16\fP) would cause background color to be displayed.
.PP
The origin of each glyph is normally set so that the character is drawn in
the upper left hand corner of the grid cell.  However, if a glyph has a
negative left bearing or an unusually large ascent, descent, or right bearing
(as is the case with \fIcursor\fP font), some character may not appear in their
own grid cells.  The \fB\-center\fP option may be used to force all glyphs to
be centered in their respective cells.
.PP
All the characters in the font may not fit in the window at once.
To see the next page of glyphs, press the \fINext\fP button at the top
of the window.  To see the previous page, press \fIPrev\fP.  To exit \fBxfd\fP,
press \fIQuit\fP.
.PP
Individual character metrics (index, width, bearings, ascent and descent) can
be displayed at the top of the window by clicking on the desired character.
.PP
The font name displayed at the top of the window is the full name of the
font, as determined by the server.  See \fBxlsfonts\fP(1) and \fBfc-list\fP(1)
for ways to generate lists of fonts, as well as more detailed summaries of
their metrics and properties.
.SH "OPTIONS"
.B xfd
accepts all of the standard toolkit command line options along with
the additional options listed below:
.TP 8
.B \-fn \fIfont\fP
This option specifies the core X server side font to be displayed.
This can also be set with
the FontGrid \fBfont\fP resource.  A font must be specified.
.TP 8
.B \-fa \fIfont\fP
This option specifies a Xft font to be displayed. This can also be set with
the FontGrid \fBface\fP resource. A font pattern must be specified.
.TP 8
.B \-box
This option indicates that a box should be displayed outlining the area
that would be filled with background color by an ImageText request.
This can also be set with
the FontGrid \fBboxChars\fP resource.  The default is False.
.TP 8
.B \-center
This option indicates that each glyph should be centered in its grid.
This can also be set with
the FontGrid \fBcenterChars\fP resource.  The default is False.
.TP 8
.B \-start \fInumber\fP
This option specifies the glyph index of the upper left hand corner of the
grid.  This is used to view characters at arbitrary locations in the font.
This can also be set with
the FontGrid \fBstartChar\fP resource.  The default is 0.
.TP 8
.B \-bc \fIcolor\fP
This option specifies the color to be used if ImageText boxes are drawn.
This can also be set with
the FontGrid \fBboxColor\fP resource.
.TP 8
.B \-rows \fInumrows\fP
This option specifies the number of rows in the grid.
This can also be set with
the FontGrid \fBcellRows\fP resource.
.TP 8
.B \-columns \fInumcols\fP
This option specifies the number of columns in the grid.
This can also be set with
the FontGrid \fBcellColumns\fP resource.
.TP 8
.B \-help
This option indicates that \fBxfd\fP should print a usage message and exit.
.TP 8
.B \-version
This option indicates that \fBxfd\fP should print version info and exit.
.SH WIDGETS
In order to specify resources, it is useful to know the
widgets which compose \fBxfd\fR.  In the notation below, indentation
indicates hierarchical structure.  The widget class name is given first,
followed by the widget instance name.
The application class name is Xfd.
.sp
.nf
.\" default T 0.8i tab stops will work here;
.\" avoids a warning during XML conversion.
Xfd  xfd
	Paned  pane
		Label  fontname
		Box  box
			Command  quit
			Command  prev
			Command  next
		Label  select
		Label  metrics
		Label  range
		Label  start
		Form  form
			FontGrid  grid
.fi
.SH FONTGRID RESOURCES
The FontGrid widget is an application-specific widget, and a subclass
of the Simple widget in the Athena widget set.  The effects and
instance names of this widget's resources are given in the
\fBOPTIONS\fP section.  Capitalize the first letter of the resource
instance name to get the corresponding class name.
.SH APPLICATION SPECIFIC RESOURCES
The instance names of the application specific resources
are given below.  Capitalize the first letter of the resource
instance name to get the corresponding class name.
These resources are unlikely to be interesting unless you are localizing
\fBxfd\fP for a different language.
.TP 8
.B selectFormat
Specifies a printf-style format string used to display information
about the selected character.  The default is "character 0x%02x%02x
(%u,%u) (%#o,%#o)".  The arguments that will come after the format string are
char.byte1, char.byte2, char.byte1, char.byte2, char.byte1, char.byte2.
char.byte1 is byte 1 of the selected character.
char.byte2 is byte 2 of the selected character.
.TP 8
.B metricsFormat
Specifies a printf-style format string used to display character
metrics.  The default is "width %d; left %d, right %d; ascent %d,
descent %d (font %d, %d)".  The arguments that will come after the
format string are the character metrics width, lbearing, rbearing,
character ascent, character descent, font ascent, and font descent.
.TP 8
.B rangeFormat
Specifies a printf-style format string used to display the range of
characters currently being displayed.  The default is "range:
0x%02x%02x (%u,%u) thru 0x%02x%02x (%u,%u)".  The arguments that will
come after the format string are the following fields from the
XFontStruct that is returned from opening the font:
min_byte1, min_char_or_byte2, min_byte1, min_char_or_byte2,
max_byte1, max_char_or_byte2, max_byte1, max_char_or_byte2.
.TP 8
.B startFormat
Specifies a printf-style format string used to display information
about the character at the upper left corner of the font grid.  The
default is "upper left: 0x%04x (%d,%d)".  The arguments that will come
after the format string are the new character, the high byte of the new
character, and the low byte of the new character.
.TP 8
.B nocharFormat
Specifies a printf-style format string to display when the selected
character does not exist.  The default is "no such character
0x%02x%02x (%u,%u) (%#o,%#o)".  The arguments that will come after the
format string are the same as for the \fBselectFormat\fP resource.
.SH "SEE ALSO"
.BR X (__miscmansuffix__),
.BR xlsfonts (__appmansuffix__),
.BR xrdb (__appmansuffix__),
.BR xfontsel (__appmansuffix__),
.BR fc-list (__appmansuffix__),
.BR fonts.conf (__filemansuffix__),
.I "X Logical Font Description Conventions"
.SH "BUGS"
The program should skip over pages full of non-existent characters.
.SH AUTHOR
Jim Fulton, MIT X Consortium; previous program of the same name by
Mark Lillibridge, MIT Project Athena.


xserver/ChangeLog
commit e597f0119cd69b6d9edf86d06d941468f90d8e6d
Author: Jeremy Huddleston <jeremyhu@apple.com>
Date:   Fri Jan 27 21:07:17 2012 -0800

    configure.ac: 1.11.4
    
    Signed-off-by: Jeremy Huddleston <jeremyhu@apple.com>

commit 6c491463ee7a3637c8aae7ddae9625b36b792078
Author: Jeremy Huddleston <jeremyhu@apple.com>
Date:   Fri Jan 20 23:54:58 2012 -0800

    1.11.3.902 (1.11.4 RC2)
    
    Signed-off-by: Jeremy Huddleston <jeremyhu@apple.com>

commit b747a91485184649c05aba6fa10734f70e926032
Author: Jeremy Huddleston <jeremyhu@apple.com>
Date:   Wed Jan 18 11:47:21 2012 -0800

    XQuartz: Bump bundle version to 2.7.1
    
    Signed-off-by: Jeremy Huddleston <jeremyhu@apple.com>

commit a1ed67fa0551ee2b39d002d3b786a2017b37c7ac
Author: Peter Hutterer <peter.hutterer@who-t.net>
Date:   Fri Jan 6 13:20:45 2012 +1000

    render: don't bother with animated cursors on floating slaves (#39989)
    
    X.Org Bug 39989 <http://bugs.freedesktop.org/show_bug.cgi?id=39989>
    
    Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>
    Reviewed-by: Jeremy Huddleston <jeremyhu@apple.com>
    (cherry picked from commit bbb6b8c834e0e1491ca14403b5d0840dd14380d3)

commit f89c39e6c9e20ffc3c1ae228ee601c2ab9007ff4
Author: Peter Hutterer <peter.hutterer@who-t.net>
Date:   Mon Jan 9 15:04:10 2012 +1000

    dix: Update pointer limits for floating devices too (#43635)
    
    When the screen is restructured, the pointer limits need to be reset for
    floating slave devices as well, not just for master pointers. Only skip
    devices that don't have a cursor (attached slaves and keyboard)
    
    Bug reproducer: float an absolute slave device, rotate the screen - the
    device is now confined to a section of the screen only.
    
    X.Org Bug 43635 <http://bugs.freedesktop.org/show_bug.cgi?id=43635>
    
    Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>
    Reviewed-by: Chase Douglas <chase.douglas@canonical.com>
    (cherry picked from commit 35bd77e9d0701daae87d681900d749604fc6471f)

commit dea9377a257a402b2c7fe4a96c958d951a7d4579
Author: Zhigang Gong <zhigang.gong@linux.intel.com>
Date:   Wed Jan 4 07:01:19 2012 +0000

    mi/mibitblt: Fix an overflow bug of bit shift.
    
    When depth equal to 32 and planeMask equal to 0, the overflow will
    occur and cause the pixmap can't be cleared. There are some test
    cases in XTS hit this bug, and this fix can eliminate the corresponding
    failures.
    
    Signed-off-by: Zhigang Gong <zhigang.gong@linux.intel.com>
    Reviewed-by: Adam Jackson <ajax@redhat.com>
    Signed-off-by: Keith Packard <keithp@keithp.com>
    (cherry picked from commit f4956faab9ccf9aba6cf9603f4489f5dad19a347)

commit d2db5100b1bafeeea8c1febd390631c8f8abb094
Author: Alan Coopersmith <alan.coopersmith@oracle.com>
Date:   Fri Oct 28 21:29:50 2011 -0700

    Convert DetermineClientCmd to use strdup instead of malloc+strncpy
    
    *cmdname is initialized to NULL earlier in the function, so it's
    okay to overwrite it with NULL if strdup fails, don't need that
    extra check.
    
    Signed-off-by: Alan Coopersmith <alan.coopersmith@oracle.com>
    Reviewed-by: Jeremy Huddleston <jeremyhu@apple.com>
    (cherry picked from commit 780133f9ae7fada462714b47e79d26075bbd9abe)

commit 5b28bcb340e279e7f4e76617e86b439092367690
Author: Adam Jackson <ajax@redhat.com>
Date:   Tue Dec 13 21:23:12 2011 -0500

    dix: Tune dixLookupDrawable for success
    
    The vast vast vast majority of resource lookups are successful.  Move some
    work to the error paths so we don't punish success.
    
    Before:
    40000000 trep @   0.0009 msec (1109091.3/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (2072652.2/sec): ShmPutImage 10x10 square
    
    After:
    40000000 trep @   0.0009 msec (1148346.9/sec): PutImage 10x10 square
    60000000 trep @   0.0005 msec (2091666.1/sec): ShmPutImage 10x10 square
    
    Reviewed-by: Alan Coopersmith <alan.coopersmith@oracle.com>
    Signed-off-by: Adam Jackson <ajax@redhat.com>
    (cherry picked from commit 8f9bdfd293ad8e45755efe8d764b4dcc2a724f51)

commit e1ea93fec93b13d18df4451210aaaf322b769b6a
Author: Adam Jackson <ajax@redhat.com>
Date:   Tue Dec 13 20:23:40 2011 -0500

    dix: Tune dtrace hooks around Dispatch
    
    Don't call LookupMajorName if the hooks aren't active, it's quite expensive.
    
    Before:
    40000000 trep @   0.0009 msec (1087458.5/sec): PutImage 10x10 square

app/xfd/xfd.c
/*
 *
 *
Copyright 1989, 1998  The Open Group

Permission to use, copy, modify, distribute, and sell this software and its
documentation for any purpose is hereby granted without fee, provided that
the above copyright notice appear in all copies and that both that
copyright notice and this permission notice appear in supporting
documentation.

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
OPEN GROUP BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN
AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

Except as contained in this notice, the name of The Open Group shall not be
used in advertising or otherwise to promote the sale, use or other dealings
in this Software without prior written authorization from The Open Group.
 * *
 * Author:  Jim Fulton, MIT X Consortium
 */

#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

#include <X11/Xfuncproto.h>
#include <X11/Intrinsic.h>
#include <X11/StringDefs.h>
#include <X11/Xos.h>
#include <X11/Xatom.h>
#include <X11/Shell.h>
#include <X11/Xaw/Cardinals.h>
#include <X11/Xaw/Paned.h>
#include <X11/Xaw/Box.h>
#include <X11/Xaw/Form.h>
#include <X11/Xaw/Command.h>
#include <stdio.h>
#include <stdlib.h>
#ifdef USE_GETTEXT
# include <X11/Xlocale.h>
# include <libintl.h>
#else
# define gettext(a) (a)
#endif
#include "grid.h"
#ifdef XRENDER
#include <X11/Xft/Xft.h>
#include <X11/extensions/Xrender.h>
#endif

static char *ProgramName;

static XrmOptionDescRec xfd_options[] = {
{"-fn",		"*grid.font",	XrmoptionSepArg,	(caddr_t) NULL },
#ifdef XRENDER
{"-fa",		"*grid.face", XrmoptionSepArg,		(caddr_t) NULL },
#endif
{"-start",	"*startChar",	XrmoptionSepArg, 	(caddr_t) NULL },
{"-box",	"*grid.boxChars", XrmoptionNoArg,	(caddr_t) "on" },
{"-bc",		"*grid.boxColor", XrmoptionSepArg, 	(caddr_t) NULL },
{"-center",	"*grid.centerChars", XrmoptionNoArg,	(caddr_t) "on" },
{"-rows",	"*grid.cellRows", XrmoptionSepArg,	(caddr_t) NULL },
{"-columns",	"*grid.cellColumns", XrmoptionSepArg,	(caddr_t) NULL },
};

static void usage(int exitval) _X_NORETURN _X_COLD;
static void SelectChar(Widget w, XtPointer closure, XtPointer data);
static void do_quit(Widget w, XEvent *event, String *params,
		    Cardinal *num_params) _X_NORETURN;
static void change_page(int page);
static void set_button_state(void);
static void do_prev(Widget w, XEvent *event, String *params,
		    Cardinal *num_params);
static void do_next(Widget w, XEvent *event, String *params,
		    Cardinal *num_params);
static void do_prev16(Widget w, XEvent *event, String *params,
		      Cardinal *num_params);
static void do_next16(Widget w, XEvent *event, String *params,
		      Cardinal *num_params);
static char *get_font_name(Display *dpy, XFontStruct *fs);
static void CatchFontConversionWarning(String name, String type, String class,
				       String defaultp, String *params,
				       Cardinal *np);

static XtActionsRec xfd_actions[] = {
  { "Quit", do_quit },
  { "Prev16", do_prev16 },
  { "Prev", do_prev },
  { "Next", do_next },
  { "Next16", do_next16 },
};

static Atom wm_delete_window;

static Widget quitButton, prev16Button, prevButton, nextButton, next16Button;


#define DEF_SELECT_FORMAT "character 0x%04x%02x (%u,%u) (%#o,%#o)"
#define DEF_METRICS_FORMAT "width %d; left %d, right %d; ascent %d, descent %d (font %d, %d)"
#define DEF_RANGE_FORMAT "range:  0x%04x%02x (%u,%u) thru 0x%04x%02x (%u,%u)"
#define DEF_START_FORMAT "upper left:  0x%06x (%d,%d)"
#define DEF_NOCHAR_FORMAT "no such character 0x%04x%02x (%u,%u) (%#o,%#o)"

static struct _xfd_resources {
  char *select_format;
  char *metrics_format;
  char *range_format;
  char *start_format;
  char *nochar_format;
} xfd_resources;

#define Offset(field) XtOffsetOf(struct _xfd_resources, field)

static XtResource Resources[] = {
  { "selectFormat", "SelectFormat", XtRString, sizeof(char *),
      Offset(select_format), XtRString, DEF_SELECT_FORMAT },
  { "metricsFormat", "MetricsFormat", XtRString, sizeof(char *),
      Offset(metrics_format), XtRString, DEF_METRICS_FORMAT },
  { "rangeFormat", "RangeFormat", XtRString, sizeof(char *),
      Offset(range_format), XtRString, DEF_RANGE_FORMAT },
  { "startFormat", "StartFormat", XtRString, sizeof(char *),
      Offset(start_format), XtRString, DEF_START_FORMAT },
  { "nocharFormat", "NocharFormat", XtRString, sizeof(char *),
      Offset(nochar_format), XtRString, DEF_NOCHAR_FORMAT },
};

#undef Offset

static void
usage(int exitval)
{
    fprintf (stderr, gettext("usage:  %s [-options ...] "), ProgramName);
    fprintf (stderr, "-fn ");
    fprintf (stderr, gettext("font\n\n"));
#ifdef XRENDER
    fprintf (stderr, gettext("        %s [-options ...] "), ProgramName);
    fprintf (stderr, "-fa ");
    fprintf (stderr, gettext("font\n\n"));
#endif
    fprintf (stderr, gettext("where options include:\n"));
    fprintf (stderr, "    -display ");
    fprintf (stderr, gettext("display       X server to contact\n"));
    fprintf (stderr, "    -geometry ");
    fprintf (stderr, gettext("geometry     size and location of window\n"));
    fprintf (stderr, "    -bc ");
    fprintf (stderr, gettext("color              color for ImageText boxes\n"));
    fprintf (stderr, "    -start ");
    fprintf (stderr, gettext("number          first character to show\n"));
    fprintf (stderr, "    -box                   ");
    fprintf (stderr, gettext("show a box around each character\n"));
    fprintf (stderr, "    -center                ");
    fprintf (stderr, gettext("center each character inside its grid\n"));
    fprintf (stderr, "    -rows ");
    fprintf (stderr, gettext("number           number of rows in grid\n"));
    fprintf (stderr, "    -columns ");
    fprintf (stderr, gettext("number        number of columns in grid\n"));
    exit (exitval);
}


static Widget selectLabel, metricsLabel, rangeLabel, startLabel, fontGrid;

static Boolean fontConversionFailed = False;
static XtErrorMsgHandler oldWarningHandler;

int
main(int argc, char *argv[])
{
    XtAppContext xtcontext;
    Widget toplevel, pane, toplabel, box, form;
    char buf[256];
    Arg av[10];
    Cardinal i;
    static XtCallbackRec cb[2] = { { SelectChar, NULL }, { NULL, NULL } };
    XFontStruct *fs;
#ifdef XRENDER
    XftFont *xft;
#endif
    char *fontname;
    long minn, maxn;

    XtSetLanguageProc(NULL, NULL, NULL);

    ProgramName = argv[0];

    /* Handle args that don't require opening a display */
    for (int n = 1; n < argc; n++) {
	const char *argn = argv[n];
	/* accept single or double dash for -help & -version */
	if (argn[0] == '-' && argn[1] == '-') {
	    argn++;
	}
	if (strcmp (argn, "-help") == 0) {
	    usage (0);
	}
	if (strcmp (argn, "-version") == 0) {
	    puts (PACKAGE_STRING);
	    exit (0);
	}
    }

    toplevel = XtAppInitialize (&xtcontext, "Xfd",
				xfd_options, XtNumber(xfd_options),
				&argc, argv, NULL, NULL, 0);

    if (argc != 1) {
	fputs(gettext("Unknown argument(s):"), stderr);
	for (int n = 1; n < argc; n++) {
	    fprintf(stderr, " %s", argv[n]);
	}
	fputs("\n\n", stderr);
	usage (1);
    }
    
#ifdef USE_GETTEXT
    {
	const char *domaindir;

	textdomain("xfd");

        /* mainly for debugging */
	if ((domaindir = getenv ("TEXTDOMAINDIR")) == NULL) {
	    domaindir = LOCALEDIR;
	}
	bindtextdomain ("xfd", domaindir);
    }
#endif

    Resources[0].default_addr = gettext(DEF_SELECT_FORMAT);
    Resources[1].default_addr = gettext(DEF_METRICS_FORMAT);
    Resources[2].default_addr = gettext(DEF_RANGE_FORMAT);
    Resources[3].default_addr = gettext(DEF_START_FORMAT);
    Resources[4].default_addr = gettext(DEF_NOCHAR_FORMAT);

    XtAppAddActions (xtcontext, xfd_actions, XtNumber (xfd_actions));
    XtOverrideTranslations
        (toplevel, XtParseTranslationTable ("<Message>WM_PROTOCOLS: Quit()"));

    XtGetApplicationResources (toplevel, (XtPointer) &xfd_resources, Resources,
			       XtNumber (Resources), NULL, ZERO);


    /* pane wrapping everything */
    pane = XtCreateManagedWidget ("pane", panedWidgetClass, toplevel,
				  NULL, ZERO);

    /* font name */
    toplabel = XtCreateManagedWidget ("fontname", labelWidgetClass, pane,
				      NULL, ZERO);

    /* button box */
    box = XtCreateManagedWidget ("box", boxWidgetClass, pane, NULL, ZERO);
    quitButton = XtCreateManagedWidget ("quit", commandWidgetClass, box,
					NULL, ZERO);
    prev16Button = XtCreateManagedWidget ("prev16", commandWidgetClass, box,
					NULL, ZERO);
    prevButton = XtCreateManagedWidget ("prev", commandWidgetClass, box,
					NULL, ZERO);
    nextButton = XtCreateManagedWidget ("next", commandWidgetClass, box,
					NULL, ZERO);
    next16Button = XtCreateManagedWidget ("next16", commandWidgetClass, box,
					NULL, ZERO);


    /* and labels in which to put information */
    selectLabel = XtCreateManagedWidget ("select", labelWidgetClass,
					 pane, NULL, ZERO);

    metricsLabel = XtCreateManagedWidget ("metrics", labelWidgetClass,
					  pane, NULL, ZERO);

    rangeLabel = XtCreateManagedWidget ("range", labelWidgetClass, pane,
					NULL, ZERO);

    startLabel = XtCreateManagedWidget ("start", labelWidgetClass, pane,
					NULL, ZERO);

    /* form in which to draw */
    form = XtCreateManagedWidget ("form", formWidgetClass, pane, NULL, ZERO);

    i = 0;
    XtSetArg (av[i], XtNtop, XtChainTop); i++;
    XtSetArg (av[i], XtNbottom, XtChainBottom); i++;
    XtSetArg (av[i], XtNleft, XtChainLeft); i++;
    XtSetArg (av[i], XtNright, XtChainRight); i++;
    XtSetArg (av[i], XtNcallback, cb); i++;

    oldWarningHandler = XtAppSetWarningMsgHandler(xtcontext,
						  CatchFontConversionWarning);

    fontGrid = XtCreateManagedWidget ("grid", fontgridWidgetClass, form,
				      av, i);

    XtAppSetWarningMsgHandler(xtcontext, oldWarningHandler);

    /* set the label at the top to tell us which font this is */
#ifdef XRENDER
    i = 0;
    XtSetArg (av[i], XtNface, &xft); i++;
    XtGetValues (fontGrid, av, i);
    if (xft)
    {
	FcChar8	*family;
	FcChar8	*style;
	FcPattern   *p;
	double	size;
	family = (FcChar8 *) "";
	FcPatternGetString (xft->pattern, FC_FAMILY, 0, &family);
	style = (FcChar8 *) "";
	FcPatternGetString (xft->pattern, FC_STYLE, 0, &style);
	size = 0;
	FcPatternGetDouble (xft->pattern, FC_SIZE, 0, &size);
	p = FcPatternBuild (NULL,
			    FC_FAMILY, FcTypeString, family,
			    FC_STYLE, FcTypeString, style,
			    FC_SIZE, FcTypeDouble, size,
			    NULL);
	fontname = (char *) FcNameUnparse (p);
	FcPatternDestroy (p);
    }
    else
#endif
    {
	i = 0;
	XtSetArg (av[i], XtNfont, &fs); i++;
	XtGetValues (fontGrid, av, i);
	if (!fs || fontConversionFailed) {
	    fprintf (stderr, gettext("%s:  no font to display\n"), ProgramName);
	    exit (1);
	}
	fontname = get_font_name (XtDisplay(toplevel), fs);
	if (!fontname) fontname = gettext("unknown font!");
    }
    i = 0;
    XtSetArg (av[i], XtNlabel, fontname); i++;
    XtSetValues (toplabel, av, i);

    minn = GridFirstChar (fontGrid);
    maxn = GridLastChar (fontGrid);
    snprintf (buf, sizeof(buf), xfd_resources.range_format,
	      minn >> 8, minn & 0xff,
	      minn >> 8, minn & 0xff,
	      maxn >> 8, maxn & 0xff,
	      maxn >> 8, maxn & 0xff);

    i = 0;
    XtSetArg (av[i], XtNlabel, buf); i++;
    XtSetValues (rangeLabel, av, i);

    XtRealizeWidget (toplevel);

    wm_delete_window = XInternAtom(XtDisplay(toplevel), "WM_DELETE_WINDOW",
                                   False);
    (void) XSetWMProtocols (XtDisplay(toplevel), XtWindow(toplevel),
                            &wm_delete_window, 1);

    change_page (0);
    XtAppMainLoop (xtcontext);
    exit(0);
}

/*ARGSUSED*/
static void
SelectChar(Widget w, XtPointer closure _X_UNUSED, XtPointer data)
{
    FontGridCharRec *p = (FontGridCharRec *) data;
    XFontStruct *fs = p->thefont;
    long n = p->thechar;
    int direction, fontascent, fontdescent;
    XCharStruct metrics;
    char buf[256];
    Arg arg;
    Boolean has_char = 1;

    XtSetArg (arg, XtNlabel, buf);

    buf[0] = '\0';
#ifdef XRENDER
    if (p->theface)
    {
	XftFont	*xft = p->theface;
	FcChar32    c = (FcChar32) n;
	has_char = (Boolean) FcCharSetHasChar (xft->charset, n);
	if (has_char)
	{
	    XGlyphInfo	extents;
	    XftTextExtents32 (XtDisplay (w), xft, &c, 1, &extents);
	    snprintf (buf, sizeof(buf), xfd_resources.metrics_format,
		      extents.xOff, - extents.x,
		      extents.xOff - extents.width + extents.x,
		      extents.y, extents.height - extents.y,
		      xft->ascent, xft->descent);
	}
    }
    else
#endif
    {
	if ((!fs->min_byte1 && !fs->max_byte1) ?
	    (n < fs->min_char_or_byte2 || n > fs->max_char_or_byte2) :
	    (n >> 8 < fs->min_byte1 || n >> 8 > fs->max_byte1 ||
	     (n & 0xff)  < fs->min_char_or_byte2 ||
	     (n & 0xff) > fs->max_char_or_byte2))
	{
	    has_char = 0;
	}
	else
	{
	    XChar2b char2b;
	    char2b.byte1 = p->thechar >> 8;
	    char2b.byte2 = p->thechar & 0xff;
	    XTextExtents16 (fs, &char2b, 1, &direction, &fontascent, &fontdescent,
			    &metrics);
	    snprintf (buf, sizeof(buf), xfd_resources.metrics_format,
		      metrics.width, metrics.lbearing, metrics.rbearing,
		      metrics.ascent, metrics.descent, fontascent, fontdescent);
	}
    }
    XtSetValues (metricsLabel, &arg, ONE);

    if (has_char)
    {
	snprintf (buf, sizeof(buf), xfd_resources.select_format,
		  n >> 8, n & 0xff,
		  n >> 8, n & 0xff,
		  n >> 8, n & 0xff);
    }
    else
    {
	snprintf (buf, sizeof(buf), xfd_resources.nochar_format,
		  n >> 8, n & 0xff,
		  n >> 8, n & 0xff,
		  n >> 8, n & 0xff);
    }
    XtSetValues (selectLabel, &arg, ONE);

    return;
}


/*ARGSUSED*/
static void
do_quit (Widget w _X_UNUSED, XEvent *event _X_UNUSED,
	 String *params _X_UNUSED, Cardinal *num_params _X_UNUSED)
{
    exit (0);
}

static void
change_page(int page)
{
    long oldstart, newstart;
    int ncols, nrows;
    Arg arg;

    arg.name = XtNstartChar;
    GetFontGridCellDimensions (fontGrid, &oldstart, &ncols, &nrows);

    if (page) {
	long start = (oldstart +
			   ((long) ncols) * ((long) nrows) * ((long) page));

	arg.value = (XtArgVal) start;
	XtSetValues (fontGrid, &arg, ONE);
    }

    /* find out what it got set to */
    arg.value = (XtArgVal) &newstart;
    XtGetValues (fontGrid, &arg, ONE);

    /* if not paging, then initialize it, else only do it actually changed */
    if (!page || newstart != oldstart) {
	unsigned int row = (unsigned int) ((newstart >> 8));
	unsigned int col = (unsigned int) (newstart & 0xff);
	char buf[256];

	XtSetArg (arg, XtNlabel, buf);
	snprintf (buf, sizeof(buf), xfd_resources.start_format,
		  newstart, row, col);
	XtSetValues (startLabel, &arg, ONE);
    }

    set_button_state ();

    return;
}


static void
set_button_state(void)
{
    Bool prevvalid, nextvalid, prev16valid, next16valid;
    Arg arg;

    GetPrevNextStates (fontGrid, &prevvalid, &nextvalid, &prev16valid, &next16valid);
    arg.name = XtNsensitive;
    arg.value = (XtArgVal) (prevvalid ? TRUE : FALSE);
    XtSetValues (prevButton, &arg, ONE);
    arg.value = (XtArgVal) (nextvalid ? TRUE : FALSE);
    XtSetValues (nextButton, &arg, ONE);
    arg.name = XtNsensitive;
    arg.value = (XtArgVal) (prev16valid ? TRUE : FALSE);
    XtSetValues (prev16Button, &arg, ONE);
    arg.value = (XtArgVal) (next16valid ? TRUE : FALSE);
    XtSetValues (next16Button, &arg, ONE);
}


/* ARGSUSED */
static void
do_prev16(Widget w _X_UNUSED, XEvent *event _X_UNUSED,
	  String *params _X_UNUSED, Cardinal *num_params _X_UNUSED)
{
    change_page (-16);
}


static void
do_prev(Widget w _X_UNUSED, XEvent *event _X_UNUSED,
	String *params _X_UNUSED, Cardinal *num_params _X_UNUSED)
{
    change_page (-1);
}


/* ARGSUSED */
static void
do_next(Widget w _X_UNUSED, XEvent *event _X_UNUSED,
	String *params _X_UNUSED, Cardinal *num_params _X_UNUSED)
{
    change_page (1);
}

/* ARGSUSED */
static void
do_next16(Widget w _X_UNUSED, XEvent *event _X_UNUSED,
	  String *params _X_UNUSED, Cardinal *num_params _X_UNUSED)
{
    change_page (16);
}


static char *
get_font_name(Display *dpy, XFontStruct *fs)
{
    register XFontProp *fp;
    register int i;
    Atom fontatom = XInternAtom (dpy, "FONT", False);

    for (i = 0, fp = fs->properties; i < fs->n_properties; i++, fp++) {
	if (fp->name == fontatom) {
	    return (XGetAtomName (dpy, fp->card32));
	}
    }
    return NULL;
}


static void
CatchFontConversionWarning(String name, String type, String class,
			   String defaultp, String *params, Cardinal *np)
{
    if (np && *np > 1 &&
	strcmp(name, "conversionError") == 0 &&
	strcmp(type, "string") == 0 &&
	strcmp(class, "XtToolkitError") == 0 &&
	strcmp(params[1], "FontStruct") == 0) fontConversionFailed = True;

    (*oldWarningHandler)(name, type, class, defaultp, params, np);
}
